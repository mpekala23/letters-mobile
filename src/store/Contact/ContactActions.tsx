import {
  Contact,
  SET_ADDING,
  SET_EXISTING,
  CLEAR_CONTACTS,
  ContactActionTypes,
} from "./ContactTypes";

export function setAdding(contact: Contact): ContactActionTypes {
  return {
    type: SET_ADDING,
    payload: contact,
  };
}

export function setExisting(contacts: Contact[]): ContactActionTypes {
  return {
    type: SET_EXISTING,
    payload: contacts,
  };
}

export function clearContacts() {
  return {
    type: CLEAR_CONTACTS,
    payload: null,
  };
}
