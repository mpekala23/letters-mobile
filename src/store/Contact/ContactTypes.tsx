import { Facility } from "types";

export const SET_ADDING = "contact/set_adding";
export const SET_EXISTING = "contact/set_existing";
export const CLEAR_CONTACTS = "contact/clear_contacts";

export interface Contact {
  id: number;
  state: string;
  firstName: string;
  lastName: string;
  inmateNumber: string;
  relationship: string;
  facility: Facility | null;
}

export interface ContactState {
  adding: Contact;
  existing: Contact[];
}

interface SetAddingAction {
  type: typeof SET_ADDING;
  payload: Contact;
}

interface SetExistingAction {
  type: typeof SET_EXISTING;
  payload: Contact[];
}

interface ClearContactsAction {
  type: typeof CLEAR_CONTACTS;
  payload: null;
}

export type ContactActionTypes =
  | SetAddingAction
  | SetExistingAction
  | ClearContactsAction;
