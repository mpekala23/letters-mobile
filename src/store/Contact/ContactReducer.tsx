import { ContactState, ContactActionTypes, SET_ADDING, SET_EXISTING } from './ContactTypes';

const initialState: ContactState = {
  adding: {
    state: '',
    first_name: '',
    last_name: '',
    inmate_number: '',
    relationship: '',
    facility: null,
  },
  existing: [],
};

export default function ContactReducer(
  state = initialState,
  action: ContactActionTypes
): ContactState {
  const currentState = { ...state };
  switch (action.type) {
    case SET_ADDING:
      currentState.adding = action.payload;
      return currentState;
    case SET_EXISTING:
      currentState.existing = action.payload;
      return currentState;
    default:
      return state;
  }
}
