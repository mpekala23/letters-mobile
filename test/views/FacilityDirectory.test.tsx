import * as React from 'react';
import { FacilityDirectoryScreen } from '@views';
import { render, toJSON, fireEvent } from '@testing-library/react-native';
import { SET_ADDING } from '@store/Contact/ContactTypes';
import { Colors } from '@styles';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

const mockStore = configureStore([]);

const setup = (facilityOverrides = {}, routeOverrides = {}) => {
  const navigation = {
    navigate: jest.fn(),
    addListener: jest.fn(),
    setParams: jest.fn(),
  };
  const route = {
    params: {
      newFacility: null,
      ...routeOverrides,
    },
  };
  const contact = {
    state: '',
    firstName: '',
    lastName: '',
    inmateNumber: '',
    relationship: '',
    facility: null,
    ...facilityOverrides,
  };
  const initialState = {
    adding: contact,
    existing: [],
  };
  const store = mockStore({
    contact: initialState,
  });

  const StoreProvider = ({ children }) => {
    return <Provider store={store}>{children}</Provider>;
  };

  return {
    navigation,
    store,
    ...render(
      <FacilityDirectoryScreen
        navigation={navigation}
        contactState={initialState}
        route={route}
      />,
      {
        wrapper: StoreProvider,
      }
    ),
  };
};

describe('Facility Directory Screen', () => {
  it('should match snapshot', () => {
    const { container } = setup();
    const tree = toJSON(container);
    expect(tree).toMatchSnapshot();
  });

  it('should have next button be disabled until facility selected', () => {
    const { navigation, getByText } = setup();
    const nextButton = getByText('Next');
    fireEvent.press(nextButton);
    expect(
      getByText('Next').parentNode.props.style[1].backgroundColor
    ).toBeDefined();
    expect(navigation.navigate).toHaveBeenCalledTimes(0);
    fireEvent.press(getByText('State Prison'));
    fireEvent.press(nextButton);
    expect(getByText('Next').parentNode.props.style[1]).toEqual({});
    expect(navigation.navigate).toHaveBeenCalledTimes(1);
  });

  it('should have the selected facility have a blue background', () => {
    const { navigation, getByText } = setup();
    const facility = getByText('State Prison').parentNode;
    expect(facility.props.style[1].backgroundColor).toBe('white');
    expect(facility.props.style[2]).toEqual({});
    fireEvent.press(facility);
    expect(facility.props.style[2].backgroundColor).toBe(Colors.SELECT);
    fireEvent.press(facility);
    expect(facility.props.style[1].backgroundColor).toBe('white');
    expect(facility.props.style[2]).toEqual({});
  });

  it('should dispatch facility info to the redux store when the next button is pressed', () => {
    const { store, getByText } = setup();
    const nextButton = getByText('Next');
    fireEvent.press(nextButton);
    expect(store.getActions().length).toBe(0);
    fireEvent.press(getByText('State Prison'));
    fireEvent.press(nextButton);
    expect(store.getActions().length).toBe(1);
    expect(store.getActions()[0]).toEqual({
      type: SET_ADDING,
      payload: {
        facility: {
          address: 'P.O. Box 400',
          city: 'Bethel',
          name: 'Yukon Kskokwim Correctional Center',
          postal: '99559',
          state: 'AK',
          type: 'State Prison',
        },
        firstName: '',
        inmateNumber: '',
        lastName: '',
        relationship: '',
        state: '',
      },
    });
  });

  it('should navigate to the contact info screen when the back button is pressed', () => {
    const { navigation, getByText } = setup();
    fireEvent.press(getByText('Back'));
    expect(navigation.navigate).toHaveBeenCalledWith('ContactInfo');
  });

  it('should navigate to the add manually screen when the add manually button is pressed', () => {
    const { navigation, getByText } = setup();
    fireEvent.press(getByText('Add Manually'));
    expect(navigation.navigate).toHaveBeenCalledWith('AddManually');
  });

  it('should navigate to the review contact screen when the next button is pressed', () => {
    const { navigation, getByText } = setup();
    const nextButton = getByText('Next');
    fireEvent.press(getByText('State Prison'));
    fireEvent.press(nextButton);
    expect(navigation.navigate).toHaveBeenCalledWith('ReviewContact');
  });

  it('should load initial facility value from the redux store', async () => {
    const { getByText } = setup({
      state: 'MN',
      firstName: 'First',
      lastName: 'Last',
      inmateNumber: '6',
      relationship: 'Sister',
      facility: {
        address: 'P.O. Box 400',
        city: 'Bethel',
        name: 'Yukon Kskokwim Correctional Center',
        postal: '99559',
        state: 'AK',
        type: 'State Prison',
      },
    });
    const facility = getByText('State Prison').parentNode;
    // need to wait so that the call to setState can be processed in the facility directory component
    setTimeout(() => {
      expect(facility.props.style[2].backgroundColor).toBe(Colors.SELECT);
    }, 1);
  });

  it('should load initial facility from manual add', async () => {
    const { navigation, getByText } = setup(
      {
        state: 'MN',
        firstName: 'First',
        lastName: 'Last',
        inmateNumber: '6',
        relationship: 'Sister',
        facility: null,
      },
      {
        newFacility: {
          address: 'P.O. Box 400',
          city: 'Bethel',
          name: 'New Yukon Kskokwim Correctional Center',
          postal: '99559',
          state: 'AK',
          type: 'State Prison',
        },
      }
    );
    const facility = getByText('New Yukon Kskokwim Correctional Center')
      .parentNode;
    // need to wait so that the call to setState can be processed in the facility directory component
    setTimeout(() => {
      expect(facility.props.style[2].backgroundColor).toBe(Colors.SELECT);
    }, 1);
    expect(navigation.setParams).toHaveBeenCalledWith({ newFacility: null });
  });
});
