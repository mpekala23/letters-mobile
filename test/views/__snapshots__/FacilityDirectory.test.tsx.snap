// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Facility Directory Screen should match snapshot 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <TouchableOpacity
    activeOpacity={1}
    style={
      Object {
        "backgroundColor": "white",
        "flex": 1,
        "paddingHorizontal": 10,
        "paddingTop": 60,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#3490dc",
            "borderBottomRightRadius": 5,
            "borderTopRightRadius": 5,
            "height": 10,
            "marginRight": 10,
            "width": 20,
          }
        }
      />
      <Text
        style={
          Object {
            "fontFamily": "Poppins-Bold",
            "fontSize": 30,
          }
        }
      >
        Facility Directory
      </Text>
    </View>
    <View
      pointerEvents="auto"
      style={
        Object {
          "backgroundColor": "#F2F2F2",
          "borderRadius": 8,
          "height": 50,
          "marginBottom": 10,
          "marginTop": 20,
          "overflow": "hidden",
          "width": "100%",
        }
      }
      testID="parent"
    >
      <View
        testID="valid"
      />
      <View
        testID="unfocused"
      />
      <ScrollView
        keyboardShouldPersistTaps="always"
        scrollEnabled={false}
        style={
          Array [
            Object {
              "flex": 1,
              "width": "100%",
            },
            Object {},
          ]
        }
      >
        <TextInput
          allowFontScaling={true}
          multiline={false}
          numberOfLines={1}
          placeholder=""
          rejectResponderTermination={true}
          secureTextEntry={false}
          style={
            Array [
              Object {
                "borderColor": "rgba(0,0,0,0)",
                "borderRadius": 8,
                "borderWidth": 2,
                "color": "#9A9A9A",
                "flex": 1,
                "fontSize": 16,
                "padding": 10,
                "width": "100%",
              },
              Object {},
              Object {},
              Object {
                "height": 50,
              },
              Object {
                "borderRadius": 20,
                "padding": 10,
              },
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
        <View />
        <View />
        <View
          style={
            Object {
              "height": 0,
              "width": "100%",
              "zIndex": 5,
            }
          }
        >
          <View />
        </View>
      </ScrollView>
    </View>
    <ScrollView
      contentContainerStyle={
        Object {
          "alignItems": "center",
          "flex": 1,
          "paddingTop": 10,
          "width": "100%",
        }
      }
      data={
        Array [
          Object {
            "address": "P.O. Box 400",
            "city": "Bethel",
            "name": "Yukon Kskokwim Correctional Center",
            "postal": "99559",
            "state": "AK",
            "type": "State Prison",
          },
        ]
      }
      disableVirtualization={false}
      horizontal={false}
      initialNumToRender={10}
      maxToRenderPerBatch={10}
      numColumns={1}
      onEndReachedThreshold={2}
      removeClippedSubviews={false}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View
        style={null}
      >
        <TouchableOpacity
          activeOpacity={0.2}
          style={
            Array [
              Object {
                "elevation": 4,
                "shadowColor": "#000",
                "shadowOffset": Object {
                  "height": 2,
                  "width": 0,
                },
                "shadowOpacity": 0.23,
                "shadowRadius": 2.62,
              },
              Object {
                "backgroundColor": "white",
                "borderRadius": 30,
                "marginBottom": 10,
                "paddingHorizontal": 30,
                "paddingVertical": 10,
                "width": 675,
              },
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Poppins-Bold",
                },
                Object {
                  "fontSize": 16,
                },
              ]
            }
          >
            Yukon Kskokwim Correctional Center
          </Text>
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Poppins-Regular-Italic",
                },
                Object {
                  "color": "#6E757C",
                },
              ]
            }
          >
            State Prison
          </Text>
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Poppins-Regular-Italic",
                },
                Object {
                  "color": "#6E757C",
                },
              ]
            }
          >
            P.O. Box 400
             -
            Bethel
            ,
            AK
             
            99559
          </Text>
        </TouchableOpacity>
      </View>
      <View>
        <View
          style={
            Object {
              "marginTop": 20,
            }
          }
        >
          <View />
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Poppins-Regular",
                },
                Object {
                  "marginBottom": 20,
                },
              ]
            }
          >
            Don't see the facility you're looking for?
          </Text>
          <TouchableOpacity
            activeOpacity={0.7}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#3490dc",
                  "borderRadius": 4,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "marginVertical": 5,
                },
                Object {},
                Object {},
                Object {},
              ]
            }
            testID="clickable"
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "white",
                    "fontSize": 20,
                    "fontWeight": "700",
                    "padding": 6,
                  },
                  Object {},
                  Object {},
                  Object {},
                ]
              }
            >
              Add Manually
            </Text>
          </TouchableOpacity>
        </View>
      </View>
    </ScrollView>
    <View
      style={
        Object {
          "alignSelf": "center",
          "flexDirection": "row",
          "marginBottom": 15,
          "marginTop": 0,
          "width": 712.5,
        }
      }
    >
      <TouchableOpacity
        activeOpacity={0.7}
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderColor": "#3490dc",
              "borderRadius": 4,
              "borderWidth": 1,
              "flexDirection": "row",
              "justifyContent": "center",
              "marginVertical": 5,
            },
            Object {},
            Object {
              "flex": 1,
              "margin": 10,
            },
            Object {},
          ]
        }
        testID="clickable"
      >
        <Text
          style={
            Array [
              Object {
                "color": "#3490dc",
                "fontSize": 20,
                "fontWeight": "700",
                "padding": 6,
              },
              Object {},
              Object {},
              Object {},
            ]
          }
        >
          Back
        </Text>
      </TouchableOpacity>
      <TouchableOpacity
        activeOpacity={1}
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#3490dc",
              "borderRadius": 4,
              "flexDirection": "row",
              "justifyContent": "center",
              "marginVertical": 5,
            },
            Object {
              "alignItems": "center",
              "backgroundColor": "grey",
              "borderRadius": 4,
              "flexDirection": "row",
              "justifyContent": "center",
              "marginVertical": 5,
            },
            Object {
              "flex": 1,
              "margin": 10,
            },
            Object {},
          ]
        }
        testID="clickable"
      >
        <Text
          style={
            Array [
              Object {
                "color": "white",
                "fontSize": 20,
                "fontWeight": "700",
                "padding": 6,
              },
              Object {
                "color": "white",
                "fontSize": 20,
                "fontWeight": "700",
                "padding": 6,
              },
              Object {},
              Object {},
            ]
          }
        >
          Next
        </Text>
      </TouchableOpacity>
    </View>
  </TouchableOpacity>
</View>
`;
